### ISSUE TITLE
`${name}-${version}-runnable.jar` is not found in deployment directory

### ISSUE THREAD

**Issue by fvarrui** [fvarrui/JavaPackager#355]
There is now a new issue. The `${name}-${version}-runnable.jar` is not found in deployment directory. In the deployement directory i find `${name}.jar` without version and 'runnable'.

Am i missing some configuration details that i have to consider in the new Version?

The very last step during packaging seems to rename the build artifact jar file.
This is the message:

```Concatenating files [/target/deploy/linux/assets/startup.sh,/target/deploy/linux/name-version-runnable.jar] into file [/target/deploy/linux/<name>.jar]```

best,
Kia

_Originally posted by @kia in https://github.com/fvarrui/JavaPackager/issues/335#issuecomment-1677088268_

**Comment by fvarrui** [comment]
> There is now a new issue. The `${name}-${version}-runnable.jar` is not found in deployment directory. In the deployement directory i find `${name}.jar` without version and 'runnable'.

As I'm seeing in your output, the runnable JAR should be in `target/deploy/linux`, as you are setting `outputDirectory` option pointing to this directory. The runnable JAR file is not there?

> Am i missing some configuration details that i have to consider in the new Version?

I think nothing has been changed.

> The very last step during packaging seems to rename the build artifact jar file. This is the message:
> 
> `Concatenating files [/target/deploy/linux/assets/startup.sh,/target/deploy/linux/name-version-runnable.jar] into file [/target/deploy/linux/<name>.jar]`

The runnable JAR is concatenated to the BASH script in order to create the executable. 

I guess you made a mistake here: `Concatenating files [/target/deploy/linux/assets/startup.sh,/target/deploy/linux/name-version-runnable.jar] into file [/target/deploy/linux/<name>]` ... the destination of concat is not a JAR file, right?

**Comment by kia** [comment]
Hi @fvarrui 

I didn't change the configuration in our pom.xml . The only thing i changed was the version number of the java packager from 1.7.2 to 1.7.3 and suddenly this error occurred. When i go back to the older version of the javapackager then it works again.

But to answer you last question: the destination of concat is an executable on linux. But then there is a `name.jar` in that directory instead of a `name-version-runnable.jar`

**Comment by kia** [comment]
In git i found that in 1.7.3 in  the class `io.github.fvarrui.javapackager.maven.CreateRunnableJar` the line (number 91) 
`element("finalName", name + "-" + version)` 
is removed in the new Version. 
This could be the cause of the problem. 
I added this line again rebuild the JavaPackager and the issue is gone.

**Comment by kia** [comment]
Hi @fvarrui,
I see the change in `maven-jar-plugin`. [Here](https://issues.apache.org/jira/browse/MJAR-233) is explained how to define the `finalName` in the correct way and if I define the `finalName` in our `pom.xml` as `${name}-${version}` then every thing works again without adding the deprecated parameter in `io.github.fvarrui.javapackager.maven.CreateRunnableJar`. 
If i leave it as `${name}` then the problem occur. 

I think the real issue is [here](https://github.com/fvarrui/JavaPackager/blob/0dd0875c389f72ca973448d816ee49b706d1e7f4/src/main/java/io/github/fvarrui/javapackager/maven/CreateRunnableJar.java#L102):
`File finalJarFile = new File(outputDirectory, name + "-" + version + "-" + classifier + ".jar");`
where you create the `fineJarFile`. May be the `finalJarFile` should be `${finalName}.jar`.

Best,
Kia

**Comment by fvarrui** [comment]
Are you setting `finalName` in your POM?

**Comment by kia** [comment]
> Are you setting `finalName` in your POM? I mean ... maybe this was causing this issue by modifying `maven-jar-plugin` behaviour, since `CreateRunnableJar` expects a different JAR filename.

yes. exactly. As soon as i set the `finalName` as `${name}-${version}` everything is fine. But I would like to set the `finalName` as `${name}`. Then there is the mismatch and the issue occur.

**Comment by fvarrui** [comment]
> yes. exactly. As soon as i set the `finalName` as `${name}-${version}` everything is fine. But I would like to set the `finalName` as `${name}`. Then there is the mismatch and the issue occur.

Ok, I guessed so ... I'm working right now on this issue. I just found how to get `finalName` programatically, so we can determine the right name of the generated runnable JAR file

**Comment by kia** [comment]
Hi @fvarrui .
I tested the snapshot `1.7.4-20230821.162917-2` and it works. I even changed the `finalName` and the JavaPackager can deal with it. 
Best,
Kia

**Comment by fvarrui** [comment]
1.7.4 released to Maven Central

**Comment by kia** [comment]
Hi @fvarrui .

the issue is unfortunately still there in the case of  `mvn clean install` . I tested the snapshot version with your fix with `mvn clean package` where it is working also in the case of `1.7.4`. 

The error message i got now is the following:

```Failed to install artifact <artifact-name>:jar:runnable:<version>: <path>/target/deploy/linux/<finalName>-runnable.jar (No such file or directory)```
where
`<finalName> = ${name}`

Again if i take `<finalName>=${name}-${version}` there is no error message and the artifact is installed.

**Comment by fvarrui** [comment]
Hi @kia!
I've just tested it and can't reproduce issue with 1.7.4:

**pom.xml**:

```xml
	<build>
		<finalName>ohlala</finalName>
		<plugins>
			<plugin>
				<groupId>io.github.fvarrui</groupId>
				<artifactId>javapackager</artifactId>
				<version>1.7.4</version>
				<executions>
					<execution>
						<phase>package</phase>
						<goals>
							<goal>package</goal>
						</goals>
						<configuration>
							<platform>windows</platform>
							<bundleJre>false</bundleJre>
							<copyDependencies>true</copyDependencies>
							<generateInstaller>false</generateInstaller>
							<outputDirectory>target/windows</outputDirectory>
							<additionalResources>
								<additionalResource>src/main/resources/info.txt</additionalResource>
								<additionalResource>HelloWorldMaven.l4j.ini</additionalResource>
							</additionalResources>
							<vmArgs>
								<vmArg>-Dcustom.variable="Hi!"</vmArg>
								<vmArg>-Dother.custom.variable="Bye!"</vmArg>
							</vmArgs>
							<fileAssociations>
								<fileAssociation>
									<description>HelloWorld File</description>
									<extension>hello</extension>
									<mimeType>application/hello</mimeType>
								</fileAssociation>
							</fileAssociations>
							<winConfig>
								<headerType>console</headerType>
								<icoFile>src/main/resources/HelloWorldMaven.ico</icoFile>
							</winConfig>
						</configuration>
					</execution>
				</executions>
			</plugin>
		</plugins>
	</build>
```

**`mvn clean package` output**:

```
[INFO]     Creating runnable JAR...
[INFO] Building jar: C:\Users\fvarrui\GitHub\HelloWorldMaven\target\windows\ohlala-runnable.jar
[INFO]         Renaming file [C:\Users\fvarrui\GitHub\HelloWorldMaven\target\windows\ohlala-runnable.jar] to [HelloWorldMaven-1.0.0-runnable.jar]
[INFO]     Runnable jar created in C:\Users\fvarrui\GitHub\HelloWorldMaven\target\windows\HelloWorldMaven-1.0.0-runnable.jar!
```

Am I missing anything?

**Comment by kia** [comment]
Did you try `mvn clean install` ? with `mvn clean package` is everything fine.

**Comment by fvarrui** [comment]
> Did you try `mvn clean install` ? with `mvn clean package` is everything fine.

Oh! 😓 sorry, I didn't read well your message, I see ... JP is renaming the runnable JAR to <name>-<version>-runnable.jar, so it fails when trying to install the original JAR into the local Maven repo

**Comment by fvarrui** [comment]
Ok, I've coded a patch, merged it into devel branch, and publish as a snapshot: `1.7.5-20230831.204758-1`

Now JP keeps the name of the runnable JAR. I could't test it exhaustively, so, please, test it and give me some feedback. Could you test it on all platforms?

**Comment by kia** [comment]
Hi @fvarrui 
Now it seems that every thin is fine. 

I tested both `mvn clean package` and `mvn clean install` on linux, mac and windows platforms. 

best,
Kia

**Comment by fvarrui** [comment]
> JP is renaming the runnable JAR to {name}-{version}-runnable.jar, so it fails when trying to install the original JAR into the local Maven repo

Great!!! And sorry, I didn't realize about this behaviour. I'm going to release JP 1.7.5 with this patch ASAP

**Comment by kia** [comment]
Thank you very much for your quick responses and fixes.

**Comment by fvarrui** [comment]
Sorry, I read your comment but forgot it. I'm going to release 1.7.5 this evening

**Comment by kia** [comment]
Thank you! :+1:

**Comment by fvarrui** [comment]
I've just released 1.7.5 to Maven Central.
I hope it works fine this time 😅 and sorry for the delay

**Comment by kia** [comment]
Hi @fvarrui 
the 1.7.5 version works.
Thank you and best regards,
Kia
